### PR TITLE
Use a sealed trait to constrain VecWithInitialized

### PR DESCRIPTION
**Pull request by vitalyd**
Use a sealed trait to constrain `VecWithInitialized`'s generic type parameter.  The safety of that struct relies on invariants upheld by `Vec<u8>`.  The struct currently has an `unsafe new` associated fn where `V: AsMut<Vec<u8>>`.
## Motivation
Avoids a bit of `unsafe`.

## Solution
This PR defines a crate private `VecU8` trait and implements it for `Vec<u8>` and `&mut Vec<u8>`.

cc @Darksonn as we discussed this a bit in https://github.com/tokio-rs/tokio/pull/3426

### CODE REVIEW

File: tokio/src/io/util/read_to_end.rs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-use crate::io::util::vec_with_initialized::{into_read_buf_parts, VecWithInitialized};
+use crate::io::util::vec_with_initialized::{into_read_buf_parts, VecWithInitialized, VecU8};
```
Comment (Darksonn) [inline tokio/src/io/util/read_to_end.rs:1]: This fixes rustfmt.
```suggestion
use crate::io::util::vec_with_initialized::{into_read_buf_parts, VecU8, VecWithInitialized};
```

File: tokio/src/io/util/read_to_end.rs [hunk 2]
```diff
@@ -31,13 +31,13 @@ where
     // SAFETY: The generic type on VecWithInitialized is &mut Vec<u8>.
```
Comment (Darksonn) [inline tokio/src/io/util/read_to_end.rs:31]:
```suggestion
```

File: tokio/src/io/util/read_to_string.rs [hunk 3]
```diff
@@ -41,7 +41,7 @@ where
     // SAFETY: The generic type of the VecWithInitialized is Vec<u8>.
```
Comment (Darksonn) [inline tokio/src/io/util/read_to_string.rs:41]:
```suggestion
```

File: tokio/src/io/util/vec_with_initialized.rs [hunk 4]
```diff
@@ -56,7 +68,7 @@ where
         self.vec.as_mut().is_empty()
     }
 
-    pub(crate) fn get_read_buf<'a>(&'a mut self) -> ReadBuf<'a> {
+    pub(crate) fn get_read_buf(&mut self) -> ReadBuf<'_> {
```
Comment (Darksonn) [inline tokio/src/io/util/vec_with_initialized.rs:71]: Can you undo this change? I'd like to keep the lifetime inside the `from_raw_parts_mut` call.

File: tokio/src/io/util/vec_with_initialized.rs [hunk 5]
```diff
@@ -1,6 +1,19 @@
...
+mod private {
+    pub trait Sealed {}
+
+    impl Sealed for Vec<u8> {}
+    impl Sealed for &mut Vec<u8> {}
+}
+
+/// A sealed trait that constrains the generic type parameter in `VecWithInitialized<V>`.  That struct's safety relies
+/// on certain invariants upheld by `Vec<u8>`.
+pub(crate) trait VecU8: AsMut<Vec<u8>> + private::Sealed {}
```
Comment (Darksonn) [inline tokio/src/io/util/vec_with_initialized.rs:13]: The question is if it is better to just make the trait unsafe instead of sealing it.

File: tokio/src/io/util/vec_with_initialized.rs (hunk 4 above)
Comment (vitalyd) [reply tokio/src/io/util/vec_with_initialized.rs:71]: Sure.  This was a bit of a gratuitous change in passing 😄 .  May I ask why you prefer the explicit `'a` though?

File: tokio/src/io/util/vec_with_initialized.rs (hunk 5 above)
Comment (vitalyd) [reply tokio/src/io/util/vec_with_initialized.rs:13]: An unsafe trait could work as well.  I felt that a sealed safe trait would be a bit better in that:
* minimizes `unsafe`s
* sealing and defining a controlled set of impls seemed pretty succinct and to the point

However, I'm open to hearing arguments for making it an unsafe trait instead.  Either way, the thing that somewhat irked me with existing code is that we weren't capturing safety via types.

File: tokio/src/io/util/vec_with_initialized.rs (hunk 4 above)
Comment (Darksonn) [reply tokio/src/io/util/vec_with_initialized.rs:71]: It's mostly that `std::slice::from_raw_parts_mut` detaches the lifetime of the resulting slice from the owner, and I try to avoid that as much as possible.

File: tokio/src/io/util/vec_with_initialized.rs (hunk 5 above)
Comment (Darksonn) [reply tokio/src/io/util/vec_with_initialized.rs:13]: I mean, either is fine. I was mostly wondering to myself if sealing even works without a crate boundary, but after some more thought, I guess it does.

File: tokio/src/io/util/vec_with_initialized.rs (hunk 4 above)
Comment (vitalyd) [reply tokio/src/io/util/vec_with_initialized.rs]: Right, `from_raw_parts_mut` uses an "unbounded" lifetime param.  However, inference should select the proper output lifetime when using the `'_` wildcard.

But, I reverted this part.

File: tokio/src/io/util/vec_with_initialized.rs (hunk 5 above)
Comment (vitalyd) [reply tokio/src/io/util/vec_with_initialized.rs:13]: Yeah, it's essentially sealing from the rest of the crate.  This allows specifying the valid types in the `vec_with_initialized` module, which is nice since it's the same source file as the uses and where the commentary about `Vec`'s invariants lives.

**Review by Darksonn — APPROVED** [review]

**Comment by Darksonn** [comment]
Thanks!
